src/service.rs

```diff
@@ -42,294 +34,163 @@ const PEER_EXPIRATION: Duration = Duration::from_secs(60);
-                    self.handle_messages(&recv_buf, (size, peer), &mut send_buf, &before_insert)
-                        .await;
-                    let now = Instant::now();
-                    let addr = peer.ip();
-                    self.peers.write().unwrap().insert(now, addr);
-                }
-            }
-        }
+    pub fn remove(&self, key: &K, timestamp: DateTime<Utc>) -> Option<V> {
+        let mut guard = self.tombstones.write().unwrap();
+        guard.insert(key.clone(), timestamp);
+
+        let ret = self.service.insert(key.clone(), (timestamp, None));
+        ret.and_then(|t| t.1)
     }
 
-    async fn start_diff_protocol(&self, send_buf: &mut Vec<u8>) {
-        let segments = {
-            let guard = self.map.read().unwrap();
-            guard.start_diff()
-        };
-        send_buf.clear();
-        for segment in segments {
-            Message::ComparisonItem::<K, V, C>(segment)
-                .serialize(&mut Serializer::new(&mut *send_buf, DefaultOptions::new()))
-                .unwrap();
-        }
-        let mut peers = self.get_peers();
-        // select a random address out of the peer network
-        // NOTE: the random address might not correspond to a real peer, so we do not add it to the
-        // list of known peers, just to our local copies of the addresses; if a peer exists at this
-        // address, they will eventually send us a message in return, and we will add them to the
-        // list of known peer
-        let addr = gen_ip(&mut *self.rng.write().unwrap(), self.peer_net);
-        peers.push(addr);
-        // initiate the reconciliation protocol with all the known peers, and a random one
-        for peer in peers {
-            trace!("start_diff {} bytes to {peer}", send_buf.len());
-            self.socket
-                .send_to(send_buf, (peer, self.port))
-                .await
-                .unwrap();
+    pub fn remove_bulk(&self, keys: &[(K, DateTime<Utc>)]) {
+        let mut guard = self.tombstones.write().unwrap();
+        for (k, t) in keys {
+            guard.insert(k.clone(), *t);
         }
+
+        self.service.insert_bulk(
+            &keys
+                .iter()
+                .map(|(k, t)| (k.clone(), (*t, None)))
+                .collect::<Vec<_>>(),
+        );
     }
 
-    async fn handle_messages<FI: Fn(&K, &V, Option<&V>)>(
-        &self,
-        recv_buf: &[u8],
-        (size, peer): (usize, SocketAddr),
-        send_buf: &mut Vec<u8>,
-        before_insert: &FI,
-    ) {
-        if size == recv_buf.len() {
-            warn!("Buffer too small for message, discarded");
-            return;
-        }
-        trace!("received {} bytes from {peer}", size);
-        let mut in_comparison = Vec::new();
-        let mut updates = Vec::new();
-        let mut deserializer = Deserializer::from_slice(&recv_buf[..size], DefaultOptions::new());
-        // read messages in buffer
+    fn handle_tombstones(&self) {
```

Comment:
I think “handle” is too generic. Maybe:
```suggestion
    fn clear_expired_tombstones(&self) {
```